config/7.3.0/fapi1part2adv/ig/scripts/groovy/ErrorResponseFilter.groovy

```diff
@@ -0,0 +1,39 @@
+/**
+ * This script is a simple implementation of HTTP Basic Authentication on
+ * server side.
+ * It expects the following arguments:
+ *  - realm: the realm to display when the user-agent prompts for
+ *    username and password if none were provided.
+ *  - username: the expected username
+ *  - password: the expected password
+ */
+SCRIPT_NAME = "[ErrorResponseFilter] - "
+
+logger.debug(SCRIPT_NAME + "Running...")
+
+return next.handle(context, request).thenOnResult {
+    response -> {
+        if(response.status.isClientError()){
+            logger.debug("{} Response has status {} - which is a Client Error", SCRIPT_NAME, response.status)
+            response.entity.getJsonAsync().then(errorResponse -> {
```

Comment:
The Promise chain here is getting lost, there's no guarantee that the logic will get executed (the .then returns a Promise but it is not being returned)

The thenOnResult needs to be changed to thenAsync, then you can return the Promise chain on the line highlighted, you just need to tweak the chain so that it returns the Response object in all cases.

I'll raise a Jira to go through the groovy in core and fix up other cases where we are not doing getJsonAsync as I've been guilty of it in the past. We can get away with it as we do not run IG in streaming mode, but in streaming mode then Promises need to be used properly to avoid blocking  / concurrency issues.

Reply:
Raised: https://github.com/SecureApiGateway/SecureApiGateway/issues/1547